src/storage/src/storage_state.rs

```diff
@@ -92,6 +96,79 @@ pub struct StorageState {
     /// Frontier of sink writes (all subsequent writes will be at times at or
     /// equal to this frontier)
     pub sink_write_frontiers: HashMap<GlobalId, Rc<RefCell<Antichain<Timestamp>>>>,
+    /// See: [SinkHandle]
+    pub sink_handles: HashMap<GlobalId, SinkHandle>,
+}
+
+/// This maintains an additional read hold on the source data for a sink, alongside
+/// the controller's hold and the handle used to read the shard internally.
+/// This is useful because environmentd's hold might expire, and the handle we use
+/// to read advances ahead of what we've successfully committed.
+/// In theory this could be stored alongside the other sink data, but this isn't
+/// intended to be a long term solution; either this should become a "critical" handle
+/// or environmentd will learn to hold its handles across restarts and this won't be
+/// needed.
+pub struct SinkHandle {
+    downgrade_tx: UnboundedSender<Antichain<Timestamp>>,
```

Comment:
what you really want here is a `tokio::sync::watch` so we don't have to reason about unbounded queues

see `src/storage/src/source/commit.rs`, basically you select on `changed()` and then clone the value out of it when it changes!